cacert: fix certificate denylist

lukegb opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

It's broken under Python 3, ironically due to the patch we're carrying.
Fix it, and add a "test" to make sure it at least builds.

Fixes #93230.

This also removes the `includeEmail` option, because it doesn't do anything.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

lukegb committed:
cacert: fix blacklist

It's broken under Python 3, ironically due to the patch we're carrying.
Fix it, and add a test to check it works.

Fixes #93230.

lukegb commented:
This will need a backport to 20.09, I think.

lukegb commented:
...and I'll need to retarget this at staging.

andir commented on pkgs/data/misc/cacert/default.nix:
```diff
@@ -66,6 +66,10 @@ stdenv.mkDerivation {
   setupHook = ./setup-hook.sh;
 
   passthru.updateScript = ./update.sh;
+  passthru.tests = [
```
Usually these are an attribute set.

lukegb replied:
Done. Opinions on this hack in general?

andir commented on pkgs/data/misc/cacert/fix-unicode-ca-names.patch:
```diff
@@ -1,20 +0,0 @@
-         continue
-     if obj['CKA_LABEL'] in blacklist:
--        print("Certificate %s blacklisted, ignoring." % obj['CKA_LABEL'])
-+        print("Certificate %s blacklisted, ignoring." % unicode(obj['CKA_LABEL']).encode('utf-8'))
```
I am guessing this patch isn't required with python3 as `str` now handles unicode (better)?

lukegb replied:
Right! This dates back to #27576 which ended up not migrating to Python 3.

andir commented on pkgs/data/misc/cacert/default.nix:
```diff
@@ -66,6 +66,12 @@ stdenv.mkDerivation {
+  passthru.tests = {
+    # Test that building this derivation with a blacklist works, and that UTF-8 is supported.
+    blacklist-utf8 = (cacert.override {
+      blacklist = [ "CFCA EV ROOT" "NetLock Arany (Class Gold) Főtanúsítvány" ];
```
Picking up the discussion started on IRC: How can we ensure that the certificates that are blacklisted are actually being execuded or in other worse that they are part of the input set.

My rough idea would be to do it like this:

```nix
{
blacklist-utf8 =
let
  blacklist = [ "CFCA EV ROOT" "NetLock Arany (Class Gold) Főtanúsítvány" ];
in runCommand "verify-the-cacert-filter-output" {
  cacertWithExludes = cacert.override { inherit blacklist; };
  inherit cacert;
} ''
  ${lib.concatStrigsSepMap "\n" (entry: "grep -q '${entry}'  $cacert || ({ echo "CA ${entry} is not part of the ca certificates (anymore). Consider picking another candidate." >&2; exit 1; })")}

  ${lib.concatStrigsSepMap "\n" (entry: "grep -v -q '${entry}'  $cacertWithExcludes || ({ echo "CA ${entry} is *still* part of the ca certificates. Our filtering code seems to be broken." >&2; exit 1; })")}
'';
}
```

lukegb committed:
cacert: remove broken includeEmail option

This doesn't do anything. Building with includeEmail = true produces
the same set as includeEmail = false, and the substitute rule removes
a random dictionary index operation.

andir commented on pkgs/data/misc/cacert/default.nix:
```diff
@@ -66,6 +70,52 @@ stdenv.mkDerivation {
+      };
+      mapBlacklist = f: concatStringsSep "\n" (mapAttrsToList f blacklistCAToFingerprint);
+    in runCommand "verify-the-cacert-filter-output" {
+      inherit openssl;
```
add `openssl` to `nativeBuildInputs` as that will probably (hopefully?) allow this to succeed even in a cross compiled build. You can then just refer to `openssl` verbatim instead of writing `$openssl/bin/openssl`.

lukegb replied:
Done. Tested with `nix-build . -A pkgsCross.raspberryPi.cacert.tests`, produced "/nix/store/z0rxlz0m6zxakkhby24h0hf0jh9v0qfh-verify-the-cacert-filter-output"

lukegb commented:
@ofborg eval
@ofborg build cacert cacert.tests

lukegb commented:
@ofborg eval
@ofborg build cacert cacert.tests

andir committed:
cacert: add myself as maintainer

lukegb commented on pkgs/data/misc/cacert/default.nix:
```diff
@@ -117,7 +117,7 @@ stdenv.mkDerivation {
     description = "A bundle of X.509 certificates of public Certificate Authorities (CA)";
     platforms = platforms.all;
-    maintainers = with maintainers; [ fpletz ];
+    maintainers = with maintainers; [ andir fpletz ];
```
I'll add myself as well, I have opinions here too it turns out...

lukegb committed:
cacert: add lukegb as maintainer

andir commented:
@lukegb welcome to the maintainers club for cacerts ;) Mind opening a PR with this change against staging-20.09?
